### ISSUE TITLE
API to listen for command execution from editor object?

### ISSUE THREAD

**Issue by kevinushey** [ajaxorg/ace#2687]
Some motivation: In RStudio, we maintain our own set of keyboard shortcuts + our own keybuffer, which we attempt to use in tandem with Ace. This means we need to do a bit of work to ensure the keybuffers are in sync; that is, if an RStudio AppCommand is fired Ace needs to clear its keybuffer, and vice versa.

Currently, it's possible to figure out if an Ace command has been executed by listening for an [`"afterExec"`](https://github.com/ajaxorg/ace/blob/aaa9e865186bab3592e5b79d0a2b8c0e540d67a8/lib/ace/commands/command_manager.js#L53) event fired from a keyboard handler; however, it is a bit unwieldy as one needs to attach to all currently registered keyboard handlers associated with a particular Ace editor instance. (for example, when using Emacs mode).

In other words, I would really like to be able to write:

```js
editor.on("commandExecuted", function(evt) { ... })
```

with the `"commandExecuted"` being executed whenever an attached keyboard handler has executed a command in response to keydown, and `evt` being a data object containing the command executed.

I think this could be implemented by adding this to the [`$callKeyboardHandlers`](https://github.com/ajaxorg/ace/blob/aaa9e865186bab3592e5b79d0a2b8c0e540d67a8/lib/ace/keyboard/keybinding.js#L101-L129) method part of the `KeyBinding` class.

Any thoughts? Or is there a simpler method of accomplishing this that I'm not aware of?

Thanks!

**Comment by nightwing** [comment]
Could you show me the relevant part of code in RStudio? In Cloud9 we didn't need to intercept $callKeyboardHandler, since it stops events that were handled by ace and they do not reach to the global handler listening on window.

Note that emacs handler doesn't execute commands, so using `editor.commands.on("afterExec", ...)`  should be enough for all keyboard handlers.

**Comment by kevinushey** [comment]
I'm not sure if it would be super helpful to show the code here (it's implemented in Java, using GWT to transpile to JavaScript, so it might be a bit opaque).

That said, the current scheme is something like this, from the RStudio side:

1. Attach an event handler that listens for `keydown` events (it gets the event before Ace does) -- this is the RStudio handler,
2. The handler will, on each `keydown`, append the key combination pressed to the current keybuffer,
3. If the keybuffer matches an RStudio command binding, execute that command, clear the keybuffer, and cancel the event (so Ace doesn't see it),
4. Otherwise, let the event pass through to Ace so that it can handle it as needed,
5. If Ace executed a command, then clear the RStudio keybuffer; otherwise, keep it alive if there exists one or more bindings with the current keybuffer as a prefix.

The main trick here is figuring out whether the 'RStudio' keybuffer needs to be cleared / refreshed after Ace sees the key -- for example, imagine a situation where an RStudio command were bound to `Ctrl+X Ctrl+P`, and an Ace command were bound to `Ctrl+X Ctrl+U`.

1. The user types `Ctrl + X` -- this doesn't match any existing RStudio command, and so passes through to Ace.
2. Ace receives the command; the command manager understands that this is a 'prefix' and so it gets handled (I think Ace handles it by seeing either `"null"` or `"chainKeys"` command),
3. RStudio gets notified of the command executed by Ace; in response, it can either clear or maintain its own keybuffer (in this case, maintain it).

This way, on the next keydown, both RStudio and Ace understand that we're looking for commands prefixed with `Ctrl + X`.

I was under the impression that separate key handlers wouldn't emit events from the `editor.commands` object, but maybe I'm wrong and listening for `"afterExec"` from `editor.commands` really is sufficient?

Either way, the over-arching goal here is -- suppose Ace receives a key down event. What's the appropriate way of figuring out whether an attached key handler actually handled the event? Does that always go through the command manager, or no?

**Comment by nightwing** [comment]
Both "afterExec" from editor.commands and "beforeEndOperation" from editor are sufficient to detect when command is executed, but they do not allow to detect "chainKeys"

`$callKeyboardHandlers` needs to emit `statusChange` event anyway https://github.com/ajaxorg/ace/pull/2690/files#diff-9689de13f4721f9f7ffd6a97abadadd8R129 to update key buffer displayed in statusbar so it can emit a more useful event instead

Should we change that line to

```js
if (success)
    this.$editor._signal("commandExecuted", toExecute);
```
?

Note that it won't be called for the textinput which is handled by https://github.com/ajaxorg/ace/pull/2690/files#diff-9689de13f4721f9f7ffd6a97abadadd8R141

Also `commandExecuted` seems to be a bad name for this since commands can be executed without keyboard interaction. maybe we should call it `keyboardActivity`?

**Comment by kevinushey** [comment]
I agree on all accounts -- emitting an event called `keyboardActivity` sounds like the right name + behaviour, and the event itself can contain the command executed (if applicable).

Huge thanks for looking into this!

**Comment by kevinushey** [comment]
What do you think about emitting `keyboardActivity` for the `textInput` case as well? In that case, the `"insertstring"` command could be emitted?

It's not strictly necessary since one can listen for the `insertstring` even from `"afterExec"` but I think it does make sense.

**Comment by nightwing** [comment]
is https://github.com/ajaxorg/ace/commit/072fc5810a8227b9e5491c1a16c59aabf6112831 better?

**Comment by kevinushey** [comment]
That looks perfect, thanks!
